Extrapolation API Basic Implmentation

Alex-Ian-Hamilton opened the pull request:
Contains implementation of the base classes for preprocessors and extrapolators (as the framework of the API) with 2 trivial implementations and the potential extrapolator example.
Also includes the 3D map base class, a basic class that sorts a 3D scalar or vector field with a header for data.

Alex-Ian-Hamilton committed:
Small comment changes.

Alex-Ian-Hamilton committed:
Merge branch 'master' of https://github.com/Alex-Ian-Hamilton/solarbextrapolation

Alex-Ian-Hamilton committed:
Updates to API inc. making extrapolators and preprocessors more similar.
Fixed (trivial) example preprocessor and extrapolator.
Tweaks to the Map3D object (save/load still not working how I want).
Early port of potential field extrapolator code (need to test output).

Alex-Ian-Hamilton committed:
Updates to the Map3D object, new format load is working.
Save works as before, using pickle.
Note: Uses a different interface to the SunPy map.

Alex-Ian-Hamilton committed:
Various updates. Mostly tweaks to the structure of the 3D map class.

Cadair commented on solarextrapolation/aMap3D.m3d:
```diff
@@ -0,0 +1,59 @@
+ccopy_reg
```
this file should be removed.

Cadair replied:
:fire:

dpshelio commented on solarextrapolation/potential_extrapolator.py:
```diff
@@ -0,0 +1,127 @@
+                    # 
+                    npmVecSpace[i,j,k,0]=-(phi[i-2,j,k]-8.0*phi[i-1,j,k]+8.0*phi[i+1,j,k]-phi[i+2,j,k])/(12.0*D)
+                    npmVecSpace[i,j,k,1]=-(phi[i,j-2,k]-8.0*phi[i,j-1,k]+8.0*phi[i,j+1,k]-phi[i,j+2,k])/(12.0*D)
+                    npmVecSpace[i,j,k,2]=-(phi[i,j,k-2]-8.0*phi[i,j,k-1]+8.0*phi[i,j,k+1]-phi[i,j,k+2])/(12.0*D)
```
cannot this be done with some kind of "binary" kernel applied to the whole image? without nested loops?  It seems like when the edge detection is done

Alex-Ian-Hamilton replied:
I'm not sure what you mean by a binary kernel?
Are you meaning an operation that can be applied to any arbitrary point in the data (array) and then using more "pythonic" syntax for implementation on the whole volume?

dpshelio commented on solarextrapolation/potential_extrapolator.py:
```diff
@@ -0,0 +1,127 @@
+        # Print the time ellapsed if debug mode.
+        if debug:
+            finish = time.time()
+            print '_determine_vec time ' + str(tupVolShape) + ': ' + str(time.strftime("%H:%M:%S", time.gmtime(finish - start)))
```
``` python
print( '_determine_vec time  {0} : {1:%H:%M:%S}'.format(tupVolShape, time.gmtime(finish - start)))
```

Alex-Ian-Hamilton replied:
Actually the time elapsed code will be redundant, as it's included in the extrapolator parent class, to provide a consistent experience on all future routines.
I'll remove it from here.

Alex-Ian-Hamilton replied:
Is this the standard format for date/time string in metadata dictionary?
I am currently using the timestamp string, I can update to this.

Alex-Ian-Hamilton committed:
Tweaks to the classes and documentation plus added unit tests.

Alex-Ian-Hamilton committed:
Changes to the potential extrapolator including tweaks to documentation, removing the elapsed time measure (redundant as it duplicates code inherited from Extrapolators) and removing all python math (library) functions (used numpy functions instead).
Working on how to use the numpy.gradient function within the potential field extrapolator. Having issues ATM.

Cadair commented on solarextrapolation/classes.py:
```diff
@@ -0,0 +1,333 @@
+    where magnetogram observations are generally taken.
+    """
+    def __init__(self, map_data, **kwargs):
+        """Method for creating a preprocessor object, using a sunpy map.
```
Style: always a line break after the first triple quotes.

Cadair commented on solarextrapolation/classes.py:
```diff
@@ -0,0 +1,333 @@
+    minimal.
+    The primary method to override is extrapolation(), the primary method to
+    call is extrapolate() which will both call extrapolation() and save the
+    result if a filepath argument is given.
```
You should document the parameters for the `__init__` function in the main docstring for the class.

i.e.

``` python

Parameters
----------

map_boundary_data : `sunpy.map.GenericMap`
    the sunpy map containing the boundary data.

...

```

Cadair commented on solarextrapolation/classes.py:
```diff
@@ -0,0 +1,333 @@
+        """
+        self.map_boundary_data = map_magnetogram
+        self.meta = { 'boundary_1_meta': self.map_boundary_data.meta }
+        self.z = kwargs.get('z', 10)
```
All these kwargs need defining in the doc string.

Cadair commented on solarextrapolation/classes.py:
```diff
@@ -0,0 +1,333 @@
+        This is the primary method to be edited in subclasses for specific
+        extrapolation routine implementations.
+        """
+        # Add some type checking, we want a map object, check for .unit attribute.
```
I think eventually these routines in the base class should do nothing, i.e. raise a `NotImplementedError` so that they have to be overriden in the subclasses.

Cadair commented on solarextrapolation/classes.py:
```diff
@@ -0,0 +1,333 @@
+        
+        return map_output
+
+    def extrapolate(self):
```
I think the `filepath` kwarg and with it the option to turn file saving on or off would be better as an argument to this function rather than the constructor, unless you have use for it else where?

Cadair commented on solarextrapolation/classes.py:
```diff
@@ -0,0 +1,333 @@
+    def __init__(self):
+        dim = 16
+        header = {'ZNAXIS': 3, 'ZNAXIS1': dim, 'ZNAXIS2': dim, 'ZNAXIS3': dim}
+        X, Y, Z = np.zeros([dim, dim, dim]), np.zeros([dim, dim, dim]), np.zeros([dim, dim, dim])
```
you could do this instead:

``` python
X, Y, Z = np.zeros([3, dim, dim, dim])
```

which would have the same effect. It unpacks the first dimension.

Cadair commented on solarextrapolation/potential_extrapolator.py:
```diff
@@ -0,0 +1,119 @@
+from classes import *
+import numpy as np
+
+class PotentialExtrapolator(Extrapolators):
```
This needs a docstring.

Cadair commented on solarextrapolation/potential_extrapolator.py:
```diff
@@ -0,0 +1,119 @@
+        phi = self._extrapolate_phi()
+        Bxyz = self._determine_vec(phi, D = 1, debug = False)
+
+        return Map3D( Bxyz, self.meta )
```
style: no space after `(`

Cadair commented on solarextrapolation/potential_extrapolator.py:
```diff
@@ -0,0 +1,119 @@
+                            # Find the components for this contribution product.
+                            B_n = boundary[i_prime, j_prime]
+                            
+                            G_n = self._Gn_5_2_29(i, j, k, i_prime, j_prime, d, d_com)
```
I wonder if you could do this using numpy ufuncs http://docs.scipy.org/doc/numpy/reference/ufuncs.html

It looks like you could, which would then vectorise it for you.

Cadair commented on solarextrapolation/potential_extrapolator.py:
```diff
@@ -0,0 +1,119 @@
+        P183 (5.2.28)
+        '''
+        if debug:
+            print 'extrapolatePhi(' + str(self.map_boundary_data.data.shape) + ', ' + str(inZ) + ', ' + str(debug) + ')'
```
instead of concatenating strings with `+` it is a  better idea to use string formatting:
i.e.

``` python
print "extrapolatePhi({},{},{})".format(self.map_boundary_data.data.shape, inZ, debug)
```

Cadair commented:
this is looking really good, once you have made the changes I will merge it in.

Just a quick note, you have to be careful with git when adding binary files, because they can bloat the repo a lot. HMI test data upto 100x100 is probably ok, but try not to change it very often (each change copies it duplicating its filesize in the history).

Alex-Ian-Hamilton committed:
Added more documentation for the potential field extrapolator.

Alex-Ian-Hamilton committed:
Added more documentation. Still having issues with sphinx maths and sphinx formatting in general.

Alex-Ian-Hamilton committed:
Additional documentions. Work on figures of merit.

Cadair commented on solarextrapolation/classes.py:
```diff
@@ -0,0 +1,391 @@
+        self.results_normalised = Table([t1, t2, t3, t4, t5, t6, t7], names=('l-infinity norm', 'test 2', 'test 3', 'test 4', 'test 5', 'test 6', 'test 7'), meta={'name': 'Results Table'})
+
+        for m in self.maps:
+            if not isinstance(m, Map3D):
```
you could replace this loop with

``` python
if not any([isinstance(m, Map3D) for m in self.maps]):
    raise ValueError(...)
```

Cadair commented:
Can you remove the `m3d` files in this PR and then we can merge it an build the docs online.

Alex-Ian-Hamilton committed:
Various changes, tweaks to documentation. Working on visulisations.
